### PR TITLE
Better error handling

### PR DESCRIPTION
**Pull request by JonathanWoollett-Light**
# Reason for This PR

1. Error propagation features excessive usage of `map_err` to wrap types before returning them.
    Some error types implement some `From` implementations, but most don't.
2. Usage of the `e` identifier is ambiguous, internally inconsistent (`e` and `err` can both be found in the code base) and externally inconsistent with typical Rust patterns (where `err` is typically used for `impl Error` types).

## Description of Changes

1. Used [`derive_more::from`](https://jeltef.github.io/derive_more/derive_more/from.html) to create many more `From` implementations for error types to allow cleaner error propagation.
   I have taken a conservative approach here only using these implementations when they can be the most simple, this leaves `map_err` still being used extensively throughout the code base. In many of these remaining instances `map_err` is used for clarity with errors where multiple variants contain the same type/s or where error variants contain multiple types.
2. Disambiguated the `e` identifier:
    - Types that implemented `Error` now use the identifier `err`, inline with the standard library and general approach
    - `Endpoint` types now use the identifier `endpoint`
    - `WriteEvent` types now use the identifier `event`

Added [`derive_more`](https://crates.io/crates/derive_more) as a dependency. It is a small (~55kb) and widely used (~17m downloads) library in the Rust ecosystem for minimizing code duplication by deriving simple implementations. Here I also restrict this further by disabling default features and only including the feature for the [`derive_more::from`](https://jeltef.github.io/derive_more/derive_more/from.html) macro.

## License Acceptance

By submitting this pull request, I confirm that my contribution is made under
the terms of the Apache 2.0 license.

## PR Checklist

- [x] All commits in this PR are signed (`git commit -s`).
- [x] The issue which led to this PR has a clear conclusion.
- [x] This PR follows the solution outlined in the related issue.
- [x] The description of changes is clear and encompassing.
- [x] Any required documentation changes (code and docs) are included in this PR.
- [x] Any newly added `unsafe` code is properly documented.
- [x] Any API changes follow the [Runbook for Firecracker API changes][2].
- [x] Any user-facing changes are mentioned in `CHANGELOG.md`.
- [x] All added/changed functionality is tested.

[1]: https://github.com/rust-vmm
[2]: ../docs/api-change-runbook.md

### CODE REVIEW

**Comment by mattschlebusch** [comment]
No objections to the objectives of the PR. Some thoughts in summary:

* Could we consider renaming `WriteEvent` to `writeEvent`. I think we lose some useful context by dropping `write` entirely from the reference name.
* The PR is already quite large. Addressing multiple standards under one review makes the investment for review larger (even if the diff size doesn't change, there is still more logic to vet for each delta).
* The code coverage delta shouldn't have happened for a coding standards change (ideally). This is likely due to the `derive_from`/`map_err` change which is the only explanation I can think of that would change the behaviour of the binary. If that's the case this is not really just a coding standards change anymore. With customer experience implications here that aren't clear (to me). I doubt it's serious, but I do think due diligence suggests checking with more experienced Firecracker Rustaceans that this is not a problem.

**Comment by JonathanWoollett-Light** [comment]
> Could we consider renaming WriteEvent to writeEvent. I think we lose some useful context by dropping write entirely from the reference name.

Rust structs typically use pascal case, however I think you mean to consider renaming the values of type `WriteEvent` to use the identifier `writeEvent`? In this case I would suggest instead `write_event` for snake case here, would you agree?

> The code coverage delta shouldn't have happened for a coding standards change (ideally). This is likely due to the derive_from/map_err change which is the only explanation I can think of that would change the behaviour of the binary. If that's the case this is not really just a coding standards change anymore. With customer experience implications here that aren't clear (to me). I doubt it's serious, but I do think due diligence suggests checking with more experienced Firecracker Rustaceans that this is not a problem.

This is what is causing the code coverage degradation here, since `derive_from` derives [`From`](https://doc.rust-lang.org/std/convert/trait.From.html) implementations for all the variants it can (some of which we may not currently use). I would note however this would not `change the behavior of the binary` as these unused implementations would not be present in the binary. Although it is not as clear as I would like (given our test suite) I do not believe this would have any affect on the Firecracker binary.

**Comment by JonathanWoollett-Light** [comment]
> With regards to the question of behaviour, is there a way we can be certain? Tests are supposed to be vetting behaviour, if the code test coverage changes surely that does indicate a possible change in behaviour?

Our tests at-least guarantee no change in most behavior we depend on, with no test failures that seems to be the case with these changes (however underlying behavior can change without necessarily changing the behavior we test).

I believe you are correct that there is some change to the binary introduced here, at the moment even combing through the PR I cannot see where this is introduced. My initial guess was new unused derived impls but I think these should be optimized out of the binary. I could look at making a diff of the assembly to see where this difference arises however this is considerable work and I do not think it is necessary given all tests pass.

I think this highlights the issue with our current code coverage testing and approach that it does not highlight the specific areas of uncovered code (we use the first approach [mentioned here](https://doc.rust-lang.org/rustc/instrument-coverage.html) rather than the second which I would consider preferable). I suggest this raises a separate issue relating to improving our code coverage process.

**Comment by mattschlebusch** [comment]
tl;dr - I'm inclined to set aside the possible coverage shortfalls in the context of this PR to move forward.

-------------------------

At the risk of being needlessly verbose, here are my passing comments:

* Code coverage tooling is lacking here and would benefit from some investment in the future. The sky is the limit in an idealistic sense, and we can consider binary-code and human-code path coverage (likely not the same). Binary coverage may be useful for vetting changes in the automated context, but it's the code coverage delta (a diff between covered lines on the base branch and covered lines of the developer branch) that would have been the most useful here for humans to scrutinize possible behavioural differences. I am also inclined to think that the `llvm-cov` implementation give the options to produce this coverage delta report.
* It's rare that code consistency is not appreciated by developers working on said code. I am still not sure though that a change this size is worthwhile given entropy in the future without programmatic enforcement of a given standard.
